src/eddymotion/utils.py

```diff
@@ -0,0 +1,166 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+#
+# Copyright 2022 The NiPreps Developers <nipreps@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+"""Utils to sort the DWI data volume indices"""
+
+import numpy as np
+from itertools import chain, zip_longest
+
+
+def linear_iterator(size=None, **kwargs):
+    """
+    Sort the DWI data volume indices linearly
+
+    Parameters
+    ----------
+    size : :obj:`int`
+        Number of volumes in the dataset
+        (for instance, the number of orientations in a DWI)
+
+    Returns
+    -------
+    :obj:`~typing.Generator`
+        The sorted index order.
+
+    Examples
+    --------
+    >>> list(linear_iterator(10))
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+    """
+    if size is None and 'bvals' in kwargs:
+        size = len(kwargs['bvals'])
+    if size is None:
+        raise TypeError("Cannot build iterator without size")
+
+    return range(size)
+
+
+def random_iterator(size=None, **kwargs):
+    """Sort the DWI data volume indices.
+
+    Parameters
+    ----------
+    size : :obj:`int`
+        Number of volumes in the dataset
+        (for instance, the number of orientations in a DWI)
+    test_seed : :obj:`bool`
+        If True, return the seed value used for the random number generator.
+
+    Returns
+    -------
+    :obj:`~typing.Generator`
+        The sorted index order.
+
+    Examples
+    --------
+    >>> random_iterator(10, seed=0, test_seed=True)
+    20210324
+    >>> random_iterator(10, seed=True, test_seed=True)
+    20210324
+    >>> random_iterator(10, seed=42, test_seed=True)
+    42
```

Comment:
What is the purpose of `test_seed`? Altering the function implementation to ease testing is backwards.

If the seed is not working, these tests will fail more often than they can possibly pass because the chances that the random sample matches the expected value of the test are really low.

```suggestion
    >>> random_iterator(15, seed=0)
    [12, 2, 7, 8, 14, 4, 9, 13, 6, 3, 5, 0, 1, 10, 11]
    >>> random_iterator(15, seed=True)
    [13, 2, 8, 12, 10, 4, 9, 3, 0, 1, 14, 6, 11, 7, 5]
    >>> random_iterator(15, seed=42)
    [0, 13, 8, 4, 7, 2, 3, 5, 10, 1, 12, 6, 11, 14, 9]
```